File: src/main/java/pl/mjedynak/idea/plugins/pit/gui/populator/ProgramParametersListPopulator.java
package pl.mjedynak.idea.plugins.pit.gui.populator;

import com.intellij.execution.configurations.ParametersList;
import pl.mjedynak.idea.plugins.pit.gui.PitConfigurationForm;

import static pl.mjedynak.idea.plugins.pit.cli.model.PitCommandLineArgument.REPORT_DIR;
import static pl.mjedynak.idea.plugins.pit.cli.model.PitCommandLineArgument.SOURCE_DIRS;
import static pl.mjedynak.idea.plugins.pit.cli.model.PitCommandLineArgument.TARGET_CLASSES;
import static pl.mjedynak.idea.plugins.pit.cli.model.PitCommandLineArgument.TARGET_TESTS;

public class ProgramParametersListPopulator {

    public void populateProgramParametersList(ParametersList programParametersList, PitConfigurationForm pitConfigurationForm) {
        addReportDir(programParametersList, pitConfigurationForm);
        addSourceDir(programParametersList, pitConfigurationForm);
        addTargetClasses(programParametersList, pitConfigurationForm);
        addTargetTests(programParametersList, pitConfigurationForm);
        addOtherParams(programParametersList, pitConfigurationForm);
    }

    private void addReportDir(ParametersList programParametersList, PitConfigurationForm pitConfigurationForm) {
        programParametersList.add(REPORT_DIR.getName());
        programParametersList.add(pitConfigurationForm.getReportDir());
    }

    private void addSourceDir(ParametersList programParametersList, PitConfigurationForm pitConfigurationForm) {
        programParametersList.add(SOURCE_DIRS.getName());
        programParametersList.add(pitConfigurationForm.getSourceDir());
    }

    private void addTargetClasses(ParametersList programParametersList, PitConfigurationForm pitConfigurationForm) {
        programParametersList.add(TARGET_CLASSES.getName());
        programParametersList.add(pitConfigurationForm.getTargetClasses());
    }

    private void addTargetTests(ParametersList programParametersList, PitConfigurationForm pitConfigurationForm) {
        /* Only add this parameter if it is set - otherwise default is the target classes */
        if (!pitConfigurationForm.getTargetTests().isEmpty()) {
            programParametersList.add(TARGET_TESTS.getName());
            programParametersList.add(pitConfigurationForm.getTargetTests());
        }
    }

    private void addOtherParams(ParametersList programParametersList, PitConfigurationForm pitConfigurationForm) {
        String otherParams = pitConfigurationForm.getOtherParams();
        String[] namesAndValues = otherParams.split(" ");
        for (String param : namesAndValues) {
            programParametersList.add(param);
        }
    }
}

File: src/main/kotlin/pl/mjedynak/idea/plugins/pit/gui/populator/ProgramParametersListPopulator.kt
package pl.mjedynak.idea.plugins.pit.gui.populator

import com.intellij.execution.configurations.ParametersList
import pl.mjedynak.idea.plugins.pit.cli.model.PitCommandLineArgument.REPORT_DIR
import pl.mjedynak.idea.plugins.pit.cli.model.PitCommandLineArgument.SOURCE_DIRS
import pl.mjedynak.idea.plugins.pit.cli.model.PitCommandLineArgument.TARGET_CLASSES
import pl.mjedynak.idea.plugins.pit.cli.model.PitCommandLineArgument.TARGET_TESTS
import pl.mjedynak.idea.plugins.pit.gui.PitConfigurationForm

class ProgramParametersListPopulator {
    fun populateProgramParametersList(
        programParametersList: ParametersList,
        pitConfigurationForm: PitConfigurationForm,
    ) {
        addReportDir(programParametersList, pitConfigurationForm)
        addSourceDir(programParametersList, pitConfigurationForm)
        addTargetClasses(programParametersList, pitConfigurationForm)
        addTargetTests(programParametersList, pitConfigurationForm)
        addOtherParams(programParametersList, pitConfigurationForm)
    }

    private fun addReportDir(
        programParametersList: ParametersList,
        pitConfigurationForm: PitConfigurationForm,
    ) {
        programParametersList.add(REPORT_DIR.argumentName)
        programParametersList.add(pitConfigurationForm.reportDir)
    }

    private fun addSourceDir(
        programParametersList: ParametersList,
        pitConfigurationForm: PitConfigurationForm,
    ) {
        programParametersList.add(SOURCE_DIRS.argumentName)
        programParametersList.add(pitConfigurationForm.sourceDir)
    }

    private fun addTargetClasses(
        programParametersList: ParametersList,
        pitConfigurationForm: PitConfigurationForm,
    ) {
        programParametersList.add(TARGET_CLASSES.argumentName)
        programParametersList.add(pitConfigurationForm.targetClasses)
    }

    private fun addTargetTests(
        programParametersList: ParametersList,
        pitConfigurationForm: PitConfigurationForm,
    ) {
        if (!pitConfigurationForm.targetTests.isEmpty()) {
            programParametersList.add(TARGET_TESTS.argumentName)
            programParametersList.add(pitConfigurationForm.targetTests)
        }
    }

    private fun addOtherParams(
        programParametersList: ParametersList,
        pitConfigurationForm: PitConfigurationForm,
    ) {
        val otherParams = pitConfigurationForm.otherParams
        val namesAndValues = otherParams.split(" ".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
        for (param in namesAndValues) {
            programParametersList.add(param)
        }
    }
}
